pkg/pma/wal.c

```diff
@@ -0,0 +1,498 @@
+                "has %zu entries\r\n",
+                wal->meta_path,
+                entry_cnt2,
+                wal->data_path,
+                entry_cnt);
+        goto close_data_file;
+    }
+
+    if (entry_cnt == 0) {
+        wal->checksum = 0;
+        if (write_all(wal->meta_fd, &wal->checksum, sizeof(wal->checksum))
+            == -1)
+        {
+            int err = errno;
+            fprintf(stderr,
+                    "wal: failed to write global checksum %llx to %s: %s\r\n",
+                    wal->checksum,
+                    wal->meta_path,
+                    strerror(err));
+            goto close_data_file;
+        }
+    } else {
+        // Verify checksums.
+        char              page[kPageIdxSz + kPageSz];
+        _metadata_entry_t entry;
+        uint64_t          checksums[]     = {0, 0};
+        uint64_t         *global_checksum = checksums;
+        uint64_t         *checksum        = checksums + 1;
+        // i is unused since read_all() implicitly updates the underlying file
+        // offset for the given file descriptor.
+        for (size_t i = 0; i < entry_cnt; i++) {
+            if (read_all(wal->data_fd, page + kPageIdxSz, kPageSz) == -1) {
+                err = errno;
+                fprintf(
+                    stderr,
+                    "wal: failed to read page #%zu from data file (%s): %s\r\n",
+                    i,
+                    wal->data_path,
+                    strerror(err));
+                goto close_data_file;
+            }
+            if (read_all(wal->meta_fd, &entry, sizeof(entry)) == -1) {
+                err = errno;
+                fprintf(
+                    stderr,
+                    "wal: failed to read entry #%zu from metadata file (%s): "
+                    "%s\r\n",
+                    i,
+                    wal->meta_path,
+                    strerror(err));
+                goto close_data_file;
+            }
+            memcpy(page, &entry.pg_idx, kPageIdxSz);
+            MurmurHash3_x86_32(page, sizeof(page), kSeed, checksum);
+            if (*checksum != entry.checksum) {
+                err = ENOTRECOVERABLE;
+                fprintf(stderr,
+                        "wal: checksum %llx computed from page #%zu of data "
+                        "file (%s) "
+                        "doesn't match checksum %llx from entry #%zu of "
+                        "metadata file "
+                        "(%s)\r\n",
+                        *checksum,
+                        i,
+                        wal->data_path,
+                        entry.checksum,
+                        i,
+                        wal->meta_path);
+                goto close_data_file;
+            }
+            // Compute the new global checksum using the existing global
+            // checksum and this entry's checksum.
+            uint64_t tmp;
+            MurmurHash3_x86_32(checksums, sizeof(checksums), kSeed, &tmp);
```

Comment:
It looks like this will only write the 4 least significant bytes of `tmp`, leaving the 4 most significant bytes uninitialized. `MurmurHash3_x86_32()` casts its last argument to `uint32_t*` before writing thru it. (The same pattern is used above, but the checksums array is zero initialized).